Make --stop-on-fail halt symex early when doing path exploration

karkhaz opened the pull request:
It's sometimes nice to make CBMC bail out as soon as it finds an unsafe path, rather than letting it continue trying to find all bugs. This is useful in the context of SV-COMP, for example.

This patchset adds a flag that enables this behaviour, including a test.

karkhaz committed:
Generalise option setting from strategy unit tests

The check_with_strategy methods previously took an argument to set the
`unwind-limit` option with. This commit replaces that argument with a
function that mutates the optionst that is passed into it. This can be
used for setting `unwind-limit` or any other option from the client.
This is in preparation for a new strategy that requires additional
options to be set.

karkhaz committed:
--stop-on-fail now stops on failed path

When doing path exploration, one sometimes wants CBMC to halt symbolic
execution as soon as it encounters a single bug, rather than continuing
to explore the program for all bugs. This behaviour can now be enabled
by passing the --stop-on-fail flag together with the --paths flag.

Prior to this commit, the --stop-on-fail flag --- which was written with
model-checking rather than path exploration in mind --- would explore
all paths, but only print out a single failed property, which isn't
terribly useful because it uses just as much time as printing out all
properties. This commit makes this flag more useful when used in
conjunction with --paths, while not changing its behaviour for model
checking.

peterschrammel commented:
There's already a `--stop-on-fail` flag. How is this one different?

karkhaz commented:
@peterschrammel thanks for pointing this out, updated the flag name, help text, and commit message to clarify.

`--stop-on-fail` still executes all paths, but stops on the first property violation. `--halt-symex-on-bug` fails much earlier: it stops exploring new paths as soon as a bug in one path is detected.

Here's the difference:
```c
int main()
{
  int x, y;
  if(x)
    y = 5 / 0;
  else
    y = 5 / 0;
}
```
```
$ cbmc --stop-on-fail --paths lifo --div-by-zero-check /tmp/foo.c
...
SAT checker: instance is SATISFIABLE
VERIFICATION FAILED
___________________________
Starting new path (1 to go)
...
SAT checker: instance is SATISFIABLE
VERIFICATION FAILED
$
$
$
$ cbmc --halt-symex-on-bug --paths lifo --div-by-zero-check /tmp/foo.c
...
** Results:
[main.division-by-zero.1] division by zero in 5 / 0: SUCCESS
[main.division-by-zero.2] division by zero in 5 / 0: FAILURE

** 1 of 2 failed (2 iterations)
VERIFICATION FAILED
```

peterschrammel commented:
> --stop-on-fail still executes all paths, but stops on the first property violation.

What does that mean? Why is this behaviour even useful?

peterschrammel commented:
> $ cbmc --halt-symex-on-bug --paths lifo --div-by-zero-check /tmp/foo.c
> ...
> ** Results:
> [main.division-by-zero.1] division by zero in 5 / 0: SUCCESS
> [main.division-by-zero.2] division by zero in 5 / 0: FAILURE

This output is confusing because `SUCCESS` doesn't mean anything here. I'd prefer either to output only the failed assertion (such as in `--stop-on-fail`) or to introduce a `UNKNOWN` status (such as in 2LS), for properties whose status hasn't been determined.

karkhaz commented:
Agreed that the current behaviour is not useful, that's what I was trying to fix with the new flag---I wasn't aware that there was an existing flag. I've updated the commit so that `--stop-on-fail` has the behaviour that I originally intended with my new flag, when used in conjunction with `--paths`. If `--paths` is not passed, then `--stop-on-fail` will just have the same behaviour as before. I've removed `--halt-symex-on-bug`, since there's no need for a new flag. Thanks!

With regards to printing assertions that weren't even checked: yes, I've been aware of this since I wrote the original path-exploration code, it's a bit annoying. I haven't got round to looking at how to fix this to be more useful. Nevertheless, the status code returned by CBMC reflects the "worst" of all paths, so at least the programatic output is correct...I will get round to fixing the printing of unexplored assertions at some point.

peterschrammel commented:
@karkhaz, thanks!

karkhaz committed:
Add clear() to path exploration worklist

Subclasses of path_storaget are now required to implement a clear()
method. This is so that if we terminate path exploration early, the
worklist can be cleared.

This commit facilitates writing tests that ensure that the worklist is
empty iff symbolic execution has finished; this should happen naturally
if symex runs its course over the entire program, but might not happen
if we terminate symex early.

This commit is in preparation for a future commit that makes the
--stop-on-fail flag terminate symbolic execution early.

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: a709135b5f28ca60a2215f43255badf96477e85d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80968200
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 93e217be9baf2d19f44be95c181b83ddd354b721).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80968304

karkhaz commented:
@peterschrammel thanks for the feedback. I've opened #2684 to make note of the spurious reporting of unchecked assertions. I've changed the title of this PR to reflect that it no longer adds a new flag. I pushed one additional commit; all tests are now passing. Cheers!

tautschnig commented on unit/path_strategies.cpp:
```diff
@@ -334,6 +373,11 @@ void _check_with_strategy(
+  if(
+    result == safety_checkert::resultt::UNSAFE &&
+    opts.get_bool_option("stop-on-fail") && opts.is_set("paths"))
+    worklist->clear();
```
Nit pick: could you please add braces around `worklist->clear();` as multi-line conditions otherwise make the code hard to read. (Applies below as well.)

tautschnig reviewed (approved):
Looks good to me except for the nit-pick.

karkhaz commented:
Fixed the nits. Does anybody have any idea why the build is broken on AWS? I force-pushed twice and it broke both times, I don't have an AWS account (perhaps I should get one)...

tautschnig commented:
@karkhaz Please rebase, you need the patch from #2716.

karkhaz commented:
Done, cheers!
